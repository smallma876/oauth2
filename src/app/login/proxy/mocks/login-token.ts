export const loginToken = {
  iss: 'https://accounts.google.com',
  nbf: 1661831589,
  aud: '374760273133-4eghe15hb1ek840nmnjotpvtelv2fq11.apps.googleusercontent.com',
  sub: '105035299156813441187',
  email: 'sergio.mallma876@gmail.com',
  email_verified: true,
  azp: '374760273133-4eghe15hb1ek840nmnjotpvtelv2fq11.apps.googleusercontent.com',
  name: 'Sergio Mallma Chavez',
  picture: 'https://lh3.googleusercontent.com/a-/AFdZucq8mJ7sluN0hMjUBWZSj0LDYWfoN4KvV-52JpyI0A=s96-c',
  given_name: 'Sergio',
  family_name: 'Mallma Chavez',
  iat: 1661831889,
  exp: 1661835489,
  jti: '677749ac9724e83aa7155eac897c2135dfcb979d',
};
