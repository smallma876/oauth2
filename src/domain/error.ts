export interface OauthError {
  message: string;
  status: number;
}
