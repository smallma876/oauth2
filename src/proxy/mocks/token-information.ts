export const TokenInformation = {
  aud: '374760273133-4eghe15hb1ek840nmnjotpvtelv2fq11.apps.googleusercontent.com',
  azp: '374760273133-4eghe15hb1ek840nmnjotpvtelv2fq11.apps.googleusercontent.com',
  email: 'sergio.mallma876@gmail.com',
  email_verified: true,
  exp: 1662353890,
  family_name: 'Mallma Chavez',
  given_name: 'Sergio',
  iat: 1662350290,
  iss: 'https://accounts.google.com',
  jti: 'ae3f781891bbad12d13c9d411ad5a45e48ccf556',
  name: 'Sergio Mallma Chavez',
  nbf: 1662349990,
  picture: 'https://lh3.googleusercontent.com/a-/AFdZucr57cAutaV8GJ70u1HWwolYmglPELDZSCrfGEFCVQ=s96-c',
  sub: '105035299156813441187',
};
