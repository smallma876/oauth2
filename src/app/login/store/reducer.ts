import produce, { Draft } from 'immer';
import { Dispatch } from 'react';
import { DispatchObject } from '../../../domain/dispatch';

export type DispatchLogin = Dispatch<DispatchObject<LoginActions>>;

export enum LoginActions {
  ClearData = 'CLEAR_DATA',
}

export interface LoginState {}

export const loginState: LoginState = {};

export const loginReducer = produce((draft: Draft<LoginState>, { type, payload }: DispatchObject<LoginActions>) => {
  switch (type) {
    case LoginActions.ClearData:
      break;
    default:
      throw new Error(`Unknown type: ${type}`);
  }
});
