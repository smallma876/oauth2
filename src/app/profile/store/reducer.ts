import produce, { Draft } from 'immer';
import { Dispatch } from 'react';
import { DispatchObject } from '../../../domain/dispatch';

export type DispatchProfile = Dispatch<DispatchObject<ProfileActions>>;

export enum ProfileActions {
  ClearData = 'CLEAR_DATA',
}

export interface ProfileState {}

export const profileState: ProfileState = {};

export const profileReducer = produce(
  (draft: Draft<ProfileState>, { type, payload }: DispatchObject<ProfileActions>) => {
    switch (type) {
      case ProfileActions.ClearData:
        break;
      default:
        throw new Error(`Unknown type: ${type}`);
    }
  },
);
