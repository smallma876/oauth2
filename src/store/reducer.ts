import produce, { Draft } from 'immer';
import { Dispatch } from 'react';
import { LoginToken } from '../app/login/domain/login-token';
import { DispatchObject } from '../domain/dispatch';

export type DispatchApp = Dispatch<DispatchObject<AppActions>>;

export enum AppActions {
  IsFetching = 'IS_FETCHING',
  FinishFetching = 'FINISHED_FETCHING',
  ErrorFetching = 'ERROR_FETCHING',
  SessionExpired = 'SESSION_EXPIRED',
  ShowLoadingBeforeNavigating = 'SHOW_LOADING_BEFORE_NAVIGATING',
  ClearError = 'CLEAR_ERROR',
  ClearData = 'CLEAR_DATA',
  SetUser = 'SET_USER',
  SetToken = 'SET_TOKEN',
}

export interface AppState {
  isFetching: boolean;
  pendingFetches: number;
  sessionExpired: boolean;
  showLoadingBeforeNavigating: boolean;
  user: LoginToken | null;
  token: string;
}

export const initialState: AppState = {
  isFetching: false,
  pendingFetches: 0,
  sessionExpired: false,
  showLoadingBeforeNavigating: false,
  user: null,
  token: '',
};

export const appReducer = produce((draft: Draft<AppState>, { type, payload }: DispatchObject<AppActions>) => {
  switch (type) {
    case AppActions.IsFetching:
      draft.pendingFetches += 1;
      draft.isFetching = true;
      break;
    case AppActions.FinishFetching:
      draft.pendingFetches -= 1;
      draft.isFetching = false;
      break;
    case AppActions.SessionExpired:
      draft.sessionExpired = true;
      break;
    case AppActions.SetUser:
      draft.user = payload as LoginToken;
      break;
    case AppActions.SetToken:
      draft.token = payload as string;
      break;
    default:
      throw new Error(`Unknown type: ${type}`);
  }
});
