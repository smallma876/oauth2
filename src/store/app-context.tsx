import { castDraft } from 'immer';
import React, { createContext, FC, ReactNode, useReducer } from 'react';
import { appReducer, AppState, DispatchApp, initialState } from './reducer';

export interface AppProviderInterface {
  children: ReactNode;
  testStateProps?: Partial<AppState>;
  testDispatch?: DispatchApp;
}

const AppStateContext = createContext<AppState | undefined>(undefined);
const AppDispatchContext = createContext<DispatchApp | undefined>(undefined);

const GlobalAppProvider: FC<AppProviderInterface> = ({
  children,
  testStateProps,
  testDispatch,
}: AppProviderInterface) => {
  const [state, dispatch] = useReducer(appReducer, initialState);

  return (
    <AppStateContext.Provider value={castDraft({ ...state, ...testStateProps })}>
      <AppDispatchContext.Provider value={testDispatch || dispatch}>{children}</AppDispatchContext.Provider>
    </AppStateContext.Provider>
  );
};

const useGlobalAppState = (): AppState => {
  const context = React.useContext(AppStateContext) as AppState;

  if (context === undefined) {
    throw new Error('useGlobalAppState must be used within AppStateContext');
  }

  return context;
};

const useGlobalAppDispatch = (): DispatchApp => {
  const context = React.useContext(AppDispatchContext) as DispatchApp;

  if (context === undefined) {
    throw new Error('useGlobalAppDispatch must be used within a AppDispatchContext');
  }

  return context;
};

export { GlobalAppProvider, useGlobalAppState, useGlobalAppDispatch };
