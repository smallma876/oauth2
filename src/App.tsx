import React, { useEffect } from 'react';
import { Route, Routes, useNavigate } from 'react-router-dom';
import Login from './app/login/pages/Home/Home';
import { LoginProvider } from './app/login/store/login-context';
import Profile from './app/profile/Profile';
import { useGlobalAppState } from './store/app-context';

function App() {
  const { user } = useGlobalAppState();
  const navigate = useNavigate();
  useEffect(() => {
    if (user) {
      navigate('/profile');
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [user]);

  return (
    <div className="App">
      <Routes>
        <Route
          path="/"
          element={
            <LoginProvider>
              <Login />
            </LoginProvider>
          }
        />
        <Route path="/profile" element={<Profile />} />
      </Routes>
    </div>
  );
}

export default App;
