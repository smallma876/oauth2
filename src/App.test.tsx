import { render, RenderResult } from '@testing-library/react';
import { act } from 'react-dom/test-utils';
import { MemoryRouter } from 'react-router-dom';
import App from './App';
import { OauthError } from './domain/error';
import { TokenInformation } from './proxy/mocks/token-information';
import { GlobalAppProvider, useGlobalAppDispatch, useGlobalAppState } from './store/app-context';
import { AppState } from './store/reducer';

interface TestComponentInterface {
  testStateProps?: Partial<AppState>;
}

const renderApp = async (initialPage: string, { testStateProps }: TestComponentInterface = {}) => {
  let container = {} as RenderResult;

  await act(async () => {
    container = render(
      <GlobalAppProvider testStateProps={testStateProps}>
        <MemoryRouter initialEntries={[initialPage]}>
          <App />
        </MemoryRouter>
      </GlobalAppProvider>,
    );
  });

  return container;
};

describe('app test', () => {
  it('should throw error that you must use useAppState inside AppStateContext', () => {
    const originalConsoleError = console.error;
    console.error = jest.fn();

    const TestComponent = () => {
      useGlobalAppState();

      return <App />;
    };

    try {
      render(<TestComponent />);
    } catch (error) {
      expect((error as OauthError).message).toBe('useGlobalAppState must be used within AppStateContext');
    }

    console.error = originalConsoleError;
  });

  it('should throw error that you must use useGlobalAppDispatch inside AppDispatchContext', () => {
    const originalConsoleError = console.error;
    console.error = jest.fn();

    const TestComponent = () => {
      useGlobalAppDispatch();

      return <App />;
    };

    try {
      render(<TestComponent />);
    } catch (error) {
      expect((error as OauthError).message).toBe('useGlobalAppDispatch must be used within a AppDispatchContext');
    }

    console.error = originalConsoleError;
  });

  it('render app', async () => {
    const { container } = await renderApp('/b');
    expect(container).toMatchSnapshot();
  });

  it('should render app with user', async () => {
    const { container } = await renderApp('/b', {
      testStateProps: {
        user: TokenInformation,
      },
    });
    expect(container).toMatchSnapshot();
  });
});
