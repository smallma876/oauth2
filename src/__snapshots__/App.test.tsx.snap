// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app test render app 1`] = `
<div>
  <div
    class="App"
  />
</div>
`;

exports[`app test should render app with user 1`] = `
<div>
  <div
    class="App"
  >
    <div
      class="sc-bczRLJ dXtvxe"
    >
      <div
        class="sc-gsnTZi fPmQmu"
      >
        <img
          alt="profile"
          class="sc-dkzDqf"
          src="https://lh3.googleusercontent.com/a-/AFdZucr57cAutaV8GJ70u1HWwolYmglPELDZSCrfGEFCVQ=s96-c"
        />
      </div>
      <div>
        Sergio Mallma Chavez
      </div>
    </div>
  </div>
</div>
`;
